File: mobile-app-uniapp/src/services/rag/rag-manager.js
/**
 * RAG Manager (移动端版本)
 *
 * 统一管理RAG检索流程
 * 集成Embeddings、VectorStore、Reranker
 *
 * 功能对齐桌面端
 */

import EmbeddingsService from './embeddings-service.js'
import VectorStore from './vector-store.js'
import Reranker from './reranker.js'
import { getHybridSearch } from './hybrid-search.js'
import { db as database } from '../database.js'

/**
 * RAG Manager类
 */
class RAGManager {
  constructor(config = {}) {
    this.config = {
      // 检索参数
      topK: config.topK || 10,
      similarityThreshold: config.similarityThreshold || 0.6,
      maxContextLength: config.maxContextLength || 6000,

      // 启用选项
      enableRAG: config.enableRAG !== false,
      enableReranking: config.enableReranking !== false,
      enableHybridSearch: config.enableHybridSearch !== false,

      // 重排序配置
      rerankMethod: config.rerankMethod || 'hybrid',
      rerankTopK: config.rerankTopK || 5,
      rerankScoreThreshold: config.rerankScoreThreshold || 0.3,

      // 权重
      vectorWeight: config.vectorWeight || 0.6,
      keywordWeight: config.keywordWeight || 0.4,

      // Embeddings配置
      embeddingsMode: config.embeddingsMode || 'auto',
      embeddingsModelName: config.embeddingsModelName || 'Xenova/all-MiniLM-L6-v2',

      ...config
    }

    // 初始化组件
    this.embeddingsService = new EmbeddingsService({
      mode: this.config.embeddingsMode,
      modelName: this.config.embeddingsModelName
    })

    this.vectorStore = new VectorStore({
      topK: this.config.topK,
      similarityThreshold: this.config.similarityThreshold
    })

    this.reranker = new Reranker({
      enabled: this.config.enableReranking,
      method: this.config.rerankMethod,
      topK: this.config.rerankTopK,
      scoreThreshold: this.config.rerankScoreThreshold,
      vectorWeight: this.config.vectorWeight,
      keywordWeight: this.config.keywordWeight
    })

    // 混合检索
    this.hybridSearch = this.config.enableHybridSearch ? getHybridSearch({
      vectorWeight: this.config.vectorWeight,
      bm25Weight: this.config.keywordWeight,
      enableCache: true,
      cacheSize: 100
    }) : null

    // 状态
    this.isInitialized = false
    this.indexBuilding = false

    // 统计
    this.stats = {
      totalQueries: 0,
      avgRetrievalTime: 0,
      cacheHits: 0,
      cacheMisses: 0
    }
  }

  /**
   * 初始化RAG Manager
   */
  async initialize() {
    console.log('[RAGManager] 初始化RAG管理器...')

    try {
      // 初始化Embeddings服务
      await this.embeddingsService.initialize()

      // 初始化VectorStore
      await this.vectorStore.initialize()

      // 构建向量索引
      await this.buildVectorIndex()

      this.isInitialized = true
      console.log('[RAGManager] ✅ RAG管理器初始化成功')

      return {
        success: true,
        mode: this.embeddingsService.currentMode,
        vectorCount: this.vectorStore.stats.totalVectors
      }
    } catch (error) {
      console.error('[RAGManager] ❌ 初始化失败:', error)
      this.isInitialized = false
      return {
        success: false,
        error: error.message
      }
    }
  }

  /**
   * 构建向量索引
   */
  async buildVectorIndex() {
    if (!this.config.enableRAG) {
      console.log('[RAGManager] RAG未启用，跳过索引构建')
      return
    }

    if (this.indexBuilding) {
      console.warn('[RAGManager] 索引正在构建中...')
      return
    }

    this.indexBuilding = true
    console.log('[RAGManager] 开始构建向量索引...')

    try {
      // 获取所有知识库项
      const items = await database.exec(`
        SELECT id, title, content, type, tags, created_at
        FROM notes
        WHERE deleted = 0
        ORDER BY updated_at DESC
        LIMIT 10000
      `)

      if (!items || items.length === 0) {
        console.log('[RAGManager] 知识库为空')
        this.indexBuilding = false
        return
      }

      console.log(`[RAGManager] 为 ${items.length} 个项目生成向量...`)

      // 批量处理
      const batchSize = 20
      let processed = 0

      for (let i = 0; i < items.length; i += batchSize) {
        const batch = items.slice(i, i + batchSize)

        try {
          // 生成嵌入向量
          const embeddings = await Promise.all(
            batch.map(async (item) => {
              const text = `${item.title}\n${item.content || ''}`
              return await this.embeddingsService.generateEmbedding(text)
            })
          )

          // 存储向量
          await this.vectorStore.addVectorsBatch(batch, embeddings)

          processed += batch.length
          console.log(`[RAGManager] 进度: ${processed}/${items.length}`)
        } catch (error) {
          console.error(`[RAGManager] 批次处理失败:`, error)
        }
      }

      // 更新IDF（用于TF-IDF）
      const allTexts = items.map(item => `${item.title}\n${item.content || ''}`)
      this.embeddingsService.updateIDF(allTexts)

      // 构建BM25索引（如果启用混合检索）
      if (this.config.enableHybridSearch && this.hybridSearch) {
        console.log('[RAGManager] 构建BM25索引...')
        await this.hybridSearch.buildBM25Index(items)
        console.log('[RAGManager] ✅ BM25索引构建完成')
      }

      console.log(`[RAGManager] ✅ 索引构建完成: ${processed}/${items.length}`)

      this.indexBuilding = false
    } catch (error) {
      console.error('[RAGManager] 索引构建失败:', error)
      this.indexBuilding = false
      throw error
    }
  }

  /**
   * 检索相关文档
   * @param {string} query - 查询文本
   * @param {Object} options - 选项
   * @returns {Promise<Array>} 检索结果
   */
  async retrieve(query, options = {}) {
    if (!this.isInitialized) {
      console.warn('[RAGManager] RAG未初始化')
      return []
    }

    const startTime = Date.now()

    try {
      let vectorResults = []

      // 1. 使用混合检索或纯向量检索
      if (this.config.enableHybridSearch && this.hybridSearch) {
        console.log('[RAGManager] 使用混合检索（向量+BM25）')

        const hybridResult = await this.hybridSearch.search(query, {
          topK: options.topK || this.config.topK,
          useCache: true,
          expandQuery: true
        })

        vectorResults = hybridResult.results

        console.log('[RAGManager] 混合检索结果:', {
          total: hybridResult.total,
          vector: hybridResult.vectorCount,
          bm25: hybridResult.bm25Count
        })
      } else {
        console.log('[RAGManager] 使用纯向量检索')

        // 生成查询向量
        const queryEmbedding = await this.embeddingsService.generateEmbedding(query)

        // 向量搜索
        vectorResults = await this.vectorStore.search(queryEmbedding, {
          topK: options.topK || this.config.topK,
          similarityThreshold: options.similarityThreshold || this.config.similarityThreshold
        })
      }

      if (vectorResults.length === 0) {
        console.log('[RAGManager] 未找到相关文档')
        return []
      }

      // 2. 重排序
      let finalResults = vectorResults

      if (this.config.enableReranking && vectorResults.length > 1) {
        finalResults = await this.reranker.rerank(query, vectorResults, {
          method: options.rerankMethod || this.config.rerankMethod,
          topK: options.rerankTopK || this.config.rerankTopK
        })
      }

      // 4. 补充完整文档信息
      const enrichedResults = await this.enrichResults(finalResults)

      // 更新统计
      const retrievalTime = Date.now() - startTime
      this.stats.totalQueries++
      this.stats.avgRetrievalTime =
        (this.stats.avgRetrievalTime * (this.stats.totalQueries - 1) + retrievalTime) /
        this.stats.totalQueries

      console.log(`[RAGManager] ✅ 检索完成: ${enrichedResults.length} 个结果，耗时 ${retrievalTime}ms`)

      return enrichedResults
    } catch (error) {
      console.error('[RAGManager] 检索失败:', error)
      return []
    }
  }

  /**
   * 增强查询（生成RAG上下文）
   * @param {string} query - 查询文本
   * @param {Object} options - 选项
   * @returns {Promise<Object>} 增强的查询
   */
  async enhanceQuery(query, options = {}) {
    const retrievedDocs = await this.retrieve(query, options)

    if (retrievedDocs.length === 0) {
      return {
        query,
        context: '',
        retrievedDocs: []
      }
    }

    // 构建上下文
    const context = retrievedDocs
      .map((doc, idx) => {
        const title = doc.title || '无标题'
        const content = doc.content || ''
        const preview = content.substring(0, 500)
        return `[文档${idx + 1}: ${title}]\n${preview}`
      })
      .join('\n\n')

    // 限制上下文长度
    const truncatedContext = context.length > this.config.maxContextLength
      ? context.substring(0, this.config.maxContextLength) + '...'
      : context

    return {
      query,
      context: truncatedContext,
      retrievedDocs: retrievedDocs.map(doc => ({
        id: doc.id,
        title: doc.title,
        content: doc.content?.substring(0, 200),
        score: doc.rerank_score || doc.similarity,
        type: doc.type
      }))
    }
  }

  /**
   * 补充完整文档信息
   */
  async enrichResults(results) {
    const enriched = []

    for (const result of results) {
      try {
        // 从数据库获取完整文档
        const doc = await database.exec(`
          SELECT id, title, content, type, tags, created_at, updated_at
          FROM notes
          WHERE id = ?
        `, [result.id])[0]

        if (doc) {
          enriched.push({
            id: doc.id,
            title: doc.title,
            content: doc.content,
            type: doc.type,
            tags: doc.tags,
            created_at: doc.created_at,
            updated_at: doc.updated_at,
            similarity: result.similarity,
            rerank_score: result.rerank_score,
            match_count: result.match_count
          })
        }
      } catch (error) {
        console.error(`[RAGManager] 补充文档失败: ${result.id}`, error)
      }
    }

    return enriched
  }

  /**
   * 添加文档到索引
   */
  async addDocument(doc) {
    try {
      const text = `${doc.title}\n${doc.content || ''}`
      const embedding = await this.embeddingsService.generateEmbedding(text)

      const metadata = {
        title: doc.title,
        content: doc.content?.substring(0, 200),
        type: doc.type,
        tags: doc.tags,
        created_at: doc.created_at
      }

      await this.vectorStore.addVector(doc.id, embedding, metadata)

      console.log(`[RAGManager] ✅ 文档已添加到索引: ${doc.id}`)
    } catch (error) {
      console.error('[RAGManager] 添加文档失败:', error)
      throw error
    }
  }

  /**
   * 从索引删除文档
   */
  async removeDocument(docId) {
    await this.vectorStore.deleteVector(docId)
    console.log(`[RAGManager] ✅ 文档已从索引删除: ${docId}`)
  }

  /**
   * 重建索引
   */
  async rebuildIndex() {
    console.log('[RAGManager] 开始重建索引...')
    await this.vectorStore.clear()
    await this.buildVectorIndex()
    console.log('[RAGManager] ✅ 索引重建完成')
  }

  /**
   * 更新配置
   */
  updateConfig(config) {
    this.config = { ...this.config, ...config }

    // 更新子组件配置
    if (config.enableReranking !== undefined) {
      this.reranker.setEnabled(config.enableReranking)
    }

    if (config.rerankMethod || config.rerankTopK || config.rerankScoreThreshold) {
      this.reranker.updateConfig({
        method: config.rerankMethod,
        topK: config.rerankTopK,
        scoreThreshold: config.rerankScoreThreshold
      })
    }

    console.log('[RAGManager] 配置已更新')
  }

  /**
   * 获取索引统计
   */
  getIndexStats() {
    return {
      enabled: this.config.enableRAG,
      initialized: this.isInitialized,
      building: this.indexBuilding,
      vectorStore: this.vectorStore.getStats(),
      embeddings: this.embeddingsService.getStats(),
      queries: {
        total: this.stats.totalQueries,
        avgTime: Math.round(this.stats.avgRetrievalTime * 100) / 100
      }
    }
  }

  /**
   * 清除缓存
   */
  clearCache() {
    this.embeddingsService.clearCache()
    console.log('[RAGManager] ✅ 缓存已清除')
  }
}

export default RAGManager

File: ios-app/ChainlessChain/Features/Knowledge/Services/RAGManager.swift
import Foundation
import CoreCommon

/// RAG Manager - Retrieval-Augmented Generation for knowledge base
/// Reference: desktop-app-vue/src/main/rag/rag-manager.js
@MainActor
class RAGManager: ObservableObject {
    static let shared = RAGManager()

    private let logger = Logger.shared

    // Services
    private let embeddingsService: EmbeddingsService
    private let vectorStore: VectorStore
    private let bm25Search: BM25Search
    private var llmManager: LLMManager
    private var knowledgeRepository: KnowledgeRepository?

    // Configuration
    @Published var config: RAGConfig

    @Published var isInitialized = false
    @Published var indexProgress: IndexProgress?

    struct RAGConfig {
        var enableRAG: Bool = true
        var topK: Int = 10
        var similarityThreshold: Float = 0.6
        var maxContextLength: Int = 6000
        var enableHybridSearch: Bool = true
        var vectorWeight: Float = 0.6
        var keywordWeight: Float = 0.4
        var chunkSize: Int = 500
        var chunkOverlap: Int = 50
    }

    struct IndexProgress {
        let processed: Int
        let total: Int
        var percentage: Int {
            total > 0 ? Int((Float(processed) / Float(total)) * 100) : 0
        }
    }

    private init() {
        self.llmManager = LLMManager.shared
        self.embeddingsService = EmbeddingsService(llmManager: llmManager)
        self.vectorStore = VectorStore(
            similarityThreshold: 0.6,
            topK: 10
        )
        self.bm25Search = BM25Search.shared
        self.config = RAGConfig()

        // Configure BM25 with default settings
        var bm25Config = BM25Config()
        bm25Config.k1 = 1.5
        bm25Config.b = 0.75
        bm25Config.language = .auto
        bm25Search.configure(bm25Config)
    }

    /// Set knowledge repository
    func setKnowledgeRepository(_ repository: KnowledgeRepository) {
        self.knowledgeRepository = repository
    }

    /// Initialize RAG manager
    func initialize() async throws {
        logger.debug("[RAGManager] Initializing RAG manager...")

        // Initialize embeddings service
        try await embeddingsService.initialize()

        // Initialize vector store
        _ = await vectorStore.initialize()

        // Build vector index
        await buildVectorIndex()

        isInitialized = true
        logger.debug("[RAGManager] RAG manager initialized successfully")
    }

    /// Build vector index from knowledge base
    func buildVectorIndex() async {
        guard config.enableRAG else {
            logger.debug("[RAGManager] RAG disabled, skipping index build")
            return
        }

        guard let repository = knowledgeRepository else {
            logger.warning("[RAGManager] Knowledge repository not set")
            return
        }

        logger.debug("[RAGManager] Building vector and BM25 index...")

        do {
            // Get all knowledge items
            let items = repository.getAllItems()

            guard !items.isEmpty else {
                logger.debug("[RAGManager] Knowledge base is empty")
                return
            }

            logger.debug("[RAGManager] Indexing \(items.count) items...")

            // Build BM25 index
            let searchDocuments = items.map { item in
                SearchDocument(
                    id: item.id,
                    content: "\(item.title)\n\(item.content ?? "")",
                    title: item.title,
                    metadata: ["type": item.type],
                    createdAt: item.createdAt
                )
            }
            bm25Search.indexDocuments(searchDocuments)
            logger.debug("[RAGManager] BM25 index built with \(items.count) documents")

            // Rebuild vector index
            try await vectorStore.rebuildIndex(items: items) { text in
                try await self.embeddingsService.generateEmbedding(text)
            }

            let stats = await vectorStore.getStats()
            logger.debug("[RAGManager] Vector index built successfully, \(stats.count) items indexed")

        } catch {
            logger.error("[RAGManager] Failed to build index: \(error)")
        }
    }

    /// Retrieve relevant knowledge for a query
    /// - Parameters:
    ///   - query: User query
    ///   - options: Retrieval options
    /// - Returns: Array of relevant documents
    func retrieve(query: String, options: RetrievalOptions? = nil) async throws -> [RetrievedDocument] {
        guard config.enableRAG else {
            return []
        }

        let opts = options ?? RetrievalOptions()
        logger.debug("[RAGManager] Retrieving for query: \"\(query)\"")

        var results: [RetrievedDocument] = []

        if opts.useHybridSearch {
            // Hybrid search: vector + keyword
            let vectorResults = try await vectorSearch(query, topK: config.topK * 2)
            let keywordResults = await keywordSearch(query, topK: config.topK * 2)

            results = mergeResults(vectorResults: vectorResults, keywordResults: keywordResults)
        } else {
            // Vector search only
            results = try await vectorSearch(query, topK: config.topK * 2)
        }

        // Deduplicate
        results = deduplicateResults(results)

        // Filter by similarity threshold
        results = results.filter { $0.score >= (opts.similarityThreshold ?? config.similarityThreshold) }

        // Limit to topK
        results = Array(results.prefix(opts.topK ?? config.topK))

        logger.debug("[RAGManager] Retrieved \(results.count) relevant items")

        return results
    }

    /// Vector search
    private func vectorSearch(_ query: String, topK: Int) async throws -> [RetrievedDocument] {
        // Generate query embedding
        let queryEmbedding = try await embeddingsService.generateEmbedding(query)

        // Search vector store
        let searchResults = try await vectorStore.search(queryEmbedding: queryEmbedding, topK: topK)

        // Convert to RetrievedDocument
        return searchResults.map { result in
            RetrievedDocument(
                id: result.id,
                title: result.metadata.title,
                content: result.document,
                type: result.metadata.type,
                score: result.score,
                source: .vector
            )
        }
    }

    /// Keyword search using BM25 algorithm
    /// BM25 provides better relevance scoring compared to simple FTS
    private func keywordSearch(_ query: String, topK: Int) async -> [RetrievedDocument] {
        // Use BM25 search for better keyword matching
        var options = SearchOptions()
        options.limit = topK

        let bm25Results = bm25Search.search(query, options: options)

        // Convert BM25 results to RetrievedDocument
        return bm25Results.map { result in
            RetrievedDocument(
                id: result.document.id,
                title: result.document.title ?? "",
                content: result.document.content,
                type: result.document.metadata["type"] ?? "text",
                score: Float(result.bm25Score),
                source: .keyword
            )
        }
    }

    /// Legacy keyword search using knowledge repository (fallback)
    private func legacyKeywordSearch(_ query: String, topK: Int) async -> [RetrievedDocument] {
        guard let repository = knowledgeRepository else {
            return []
        }

        // Search knowledge base using FTS
        let items = repository.searchItems(query: query, limit: topK)

        // Convert to RetrievedDocument with estimated score
        return items.map { item in
            RetrievedDocument(
                id: item.id,
                title: item.title,
                content: item.content ?? "",
                type: item.type,
                score: 0.5, // FTS doesn't provide scores, use default
                source: .keyword
            )
        }
    }

    /// Merge vector and keyword search results using RRF (Reciprocal Rank Fusion)
    /// RRF formula: score = Σ weight / (k + rank)
    private func mergeResults(vectorResults: [RetrievedDocument], keywordResults: [RetrievedDocument]) -> [RetrievedDocument] {
        let k: Float = 60.0  // RRF parameter
        var scoreMap: [String: Float] = [:]
        var docMap: [String: RetrievedDocument] = [:]

        // Vector Search RRF scores
        for (rank, result) in vectorResults.enumerated() {
            let rrfScore = config.vectorWeight / (k + Float(rank + 1))
            scoreMap[result.id, default: 0] += rrfScore

            var doc = result
            doc.vectorScore = result.score
            docMap[result.id] = doc
        }

        // BM25 Search RRF scores
        for (rank, result) in keywordResults.enumerated() {
            let rrfScore = config.keywordWeight / (k + Float(rank + 1))
            scoreMap[result.id, default: 0] += rrfScore

            if var existing = docMap[result.id] {
                existing.keywordScore = result.score
                existing.source = .hybrid
                docMap[result.id] = existing
            } else {
                var doc = result
                doc.keywordScore = result.score
                docMap[result.id] = doc
            }
        }

        // Build final results with RRF scores
        var merged: [RetrievedDocument] = []
        for (docId, rrfScore) in scoreMap {
            guard var doc = docMap[docId] else { continue }
            doc.score = rrfScore
            merged.append(doc)
        }

        // Sort by RRF score descending
        return merged.sorted { $0.score > $1.score }
    }

    /// Deduplicate results by ID, keeping higher score
    private func deduplicateResults(_ results: [RetrievedDocument]) -> [RetrievedDocument] {
        var seen: [String: RetrievedDocument] = [:]

        for result in results {
            if let existing = seen[result.id] {
                if result.score > existing.score {
                    seen[result.id] = result
                }
            } else {
                seen[result.id] = result
            }
        }

        return seen.values.sorted { $0.score > $1.score }
    }

    /// Build enhanced context for LLM from retrieved documents
    func buildEnhancedContext(query: String, retrievedDocs: [RetrievedDocument]) -> String {
        guard !retrievedDocs.isEmpty else {
            return ""
        }

        var context = "# 相关知识库内容\n\n"
        var currentLength = context.count

        for doc in retrievedDocs {
            let docText = "## \(doc.title)\n\(doc.content)\n\n"

            // Check length limit
            if currentLength + docText.count > config.maxContextLength {
                break
            }

            context += docText
            currentLength += docText.count
        }

        context += "\n# 用户问题\n\(query)\n\n请基于以上知识库内容回答用户问题。"

        return context
    }

    /// RAG-enhanced query
    func enhanceQuery(query: String, options: RetrievalOptions? = nil) async throws -> EnhancedQuery {
        guard config.enableRAG else {
            return EnhancedQuery(query: query, context: "", retrievedDocs: [])
        }

        do {
            let retrievedDocs = try await retrieve(query: query, options: options)
            let context = buildEnhancedContext(query: query, retrievedDocs: retrievedDocs)

            return EnhancedQuery(
                query: query,
                context: context,
                retrievedDocs: retrievedDocs
            )
        } catch {
            logger.error("[RAGManager] Failed to enhance query: \(error)")
            return EnhancedQuery(query: query, context: "", retrievedDocs: [])
        }
    }

    /// Add item to vector and BM25 index
    func addToIndex(_ item: KnowledgeItem) async throws {
        guard config.enableRAG else { return }

        let text = "\(item.title)\n\(item.content ?? "")"

        // Add to BM25 index
        let searchDoc = SearchDocument(
            id: item.id,
            content: text,
            title: item.title,
            metadata: ["type": item.type],
            createdAt: item.createdAt
        )
        bm25Search.addDocument(searchDoc)

        // Add to vector index
        let embedding = try await embeddingsService.generateEmbedding(text)

        let metadata = VectorStore.VectorMetadata(
            title: item.title,
            type: item.type,
            createdAt: item.createdAt,
            updatedAt: item.updatedAt
        )

        try await vectorStore.addVector(
            id: item.id,
            embedding: embedding,
            metadata: metadata,
            document: item.content ?? item.title
        )

        logger.debug("[RAGManager] Added item to index: \(item.id)")
    }

    /// Remove item from vector and BM25 index
    func removeFromIndex(_ itemId: String) async throws {
        // Remove from BM25 index
        bm25Search.removeDocument(itemId)

        // Remove from vector index
        try await vectorStore.deleteVector(id: itemId)

        logger.debug("[RAGManager] Removed item from index: \(itemId)")
    }

    /// Update item in index
    func updateIndex(_ item: KnowledgeItem) async throws {
        try await removeFromIndex(item.id)
        try await addToIndex(item)
    }

    /// Rebuild entire index
    func rebuildIndex() async throws {
        logger.debug("[RAGManager] Rebuilding index...")

        // Clear BM25 index
        bm25Search.clear()

        // Clear embedding cache
        embeddingsService.clearCache()

        // Rebuild both indexes
        await buildVectorIndex()
    }

    /// Get index statistics
    func getIndexStats() async -> IndexStats {
        let vectorStats = await vectorStore.getStats()
        let cacheStats = embeddingsService.getCacheStats()
        let bm25Stats = bm25Search.getStats()

        return IndexStats(
            totalItems: vectorStats.count,
            storageMode: vectorStats.mode,
            cacheHitRate: cacheStats.hitRate,
            cacheSize: cacheStats.size,
            bm25DocumentCount: bm25Stats.documentCount,
            bm25AvgDocLength: bm25Stats.avgDocLength
        )
    }

    struct IndexStats {
        let totalItems: Int
        let storageMode: String
        let cacheHitRate: Double
        let cacheSize: Int
        let bm25DocumentCount: Int
        let bm25AvgDocLength: Double
    }
}

// MARK: - Data Types

struct RetrievalOptions {
    var topK: Int?
    var similarityThreshold: Float?
    var useHybridSearch: Bool = true
}

struct RetrievedDocument: Identifiable {
    let id: String
    let title: String
    let content: String
    let type: String
    var score: Float
    var source: SearchSource
    var vectorScore: Float = 0
    var keywordScore: Float = 0
}

enum SearchSource {
    case vector
    case keyword
    case hybrid
}

struct EnhancedQuery {
    let query: String
    let context: String
    let retrievedDocs: [RetrievedDocument]
}

